include/wil/cppwinrt_register_com_server.h

```diff
@@ -0,0 +1,90 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+//! @file
+//! Utilities for making managing COM server easier
+
+#ifndef __WIL_CPPWINRT_REGISTER_COM_SERVER_DEFINED
+#define __WIL_CPPWINRT_REGISTER_COM_SERVER_DEFINED
+
+#if _MSVC_LANG >= 201703L
+
+#include <winrt/base.h>
+#include <vector>
+
+namespace wil::details
+{
+template <typename T>
+struct Factory : winrt::implements<Factory<T>, IClassFactory, winrt::no_module_lock>
+{
+    HRESULT __stdcall CreateInstance(IUnknown* outer, GUID const& iid, void** result) noexcept final
+    {
+        *result = nullptr;
+
+        if (outer)
+        {
+            return CLASS_E_NOAGGREGATION;
+        }
+
+        return winrt::make_self<T>().as(iid, result);
+    }
+
+    HRESULT __stdcall LockServer(BOOL) noexcept final
+    {
+        return S_OK;
+    }
+};
+
+template <typename T = void, typename... Rest>
+void register_com_server(std::vector<DWORD>& registrations)
+{
+    auto& back = registrations.emplace_back();
+    winrt::check_hresult(
+        CoRegisterClassObject(winrt::guid_of<T>(), winrt::make<Factory<T>>().get(), CLSCTX_LOCAL_SERVER, REGCLS_MULTIPLEUSE, &back));
+    register_com_server<Rest...>(registrations);
+}
+
+template <>
+void register_com_server<void>(std::vector<DWORD>&)
+{
+}
+} // namespace wil::details
+
+namespace wil
+{
+struct com_server_revoker
+{
+    std::vector<DWORD> registrations;
+    ~com_server_revoker()
+    {
+        revoke();
+    }
+    void revoke()
+    {
+        for (auto&& registration : registrations)
+        {
+            winrt::check_hresult(CoRevokeClassObject(registration));
+        }
+    }
+};
+
+template <typename T, typename... Rest>
+[[nodiscard]] com_server_revoker register_com_server()
+{
+    std::vector<DWORD> registrations;
+    registrations.reserve(sizeof...(Rest) + 1);
+    details::register_com_server<T, Rest...>(registrations);

```

Comment:
If this throws in the middle of registration, you'll leak all previous registrations

Reply:
Does WRL's [Module::Registers](https://learn.microsoft.com/en-us/cpp/cppcx/wrl/module-class?view=msvc-170#registerobjects) fail-fast for any error, or just skip classes that failed to register? I'd like to imitate the behavior here as the ask is "WRL but in cppwinrt speak".

Reply:
Updated. Follow WRL to undo all registered classes upon any failures. Class revocation errors are ignored as per WRL.